better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+array of pairs, but stores the elements of those pairs in a pair of
+arrays.[^array-pairs]
+
+[^array-pairs]: You might want to use a type like this one to store
+    `(Bool, Int64)` pairs without wasting storage for padding, or to
+    leverage SIMD operations.
+
+```swift
+/// A series of `(X, Y)` pairs.
+struct PairArray<X, Y> {
+  // Invariant: `xs.count == ys.count`
+
+  /// The first part of each element.
+  private var xs: [X] = []
+
+  /// The second part of each element.
+  private var ys: [Y] = []
+
+  /// An empty instance.
+  public init() {}
+
+  /// The `i`th element.
+  public subscript(i: Int) -> (X, Y) {
+    (xs[i], ys[i])
+  }
+
+  /// The length.
+  public var count: Int { ys.count }
+
+  /// Adds `e` to the end.
+  public mutating func append(_ e: (X, Y)) {
+    xs.append(e.0)
+    ys.append(e.1)
+  }
+}
+```
+
+The invariant for this type is that the private arrays have the same
+length.  It's important to remember that invariants only hold at a
+type's public interface boundary and are routinely violated,
+temporarily, durign a mutation.  For example, in `append`, we have to
+grow one of the arrays first, which breaks the invariant until we've
+done the other push_back.  That's not a problem because the arrays are
+private—that “bad” state is *encapsulated* by the type, and
+invisible to clients.  By the time we return from `append`, everything
+is back in order.
+
+Because we can control access to the visibility of “bad” states, there
+is a formula for upholding a type invariant.  We
+
+- make stored properties `private` or `private(set)`
+- establish the invariant in the type's `init` method(s), and
+- in `mutating` methods and in the mutation clauses of subscripts
+  and computed properties, ensure the invariant is intact:
+  - before returning to the caller
+  - before passing `self` as a parameter
+  - before calling a non-private method or subscript—or accessing a non-private computed property—of `self`.
+
+Non-mutating operations can't disturb the invariant, and any mutating
+operation that only uses the type's public API trivially upholds the
+invariant.
+
+#### How To Choose a Type Invariant
+
+Often, you have a choice about how strong to make your type's
+invariant.  For example, in `PairArray`, we could have chosen the
+invariant `x.count <= y.s.count`.  We call that a *weaker* invariant
+because it allows many more internal states.
+
+Let's assume that there's some method that can create a condition
+where `x.count < y.count`:
+
+```swift
+  /// An instance with the value of `zip(xs, ys)`.
+  ///
+  /// - Precondition: `xs.count <= ys.count`.
+  init(xs: [X], ys: [Y]) { (self.xs, self.ys) = (xs, ys) }
+```
+
+[Note: when sequences of unequal length are `zip`ped, the result
+has the length of the shorter sequence.]
+
+Now let's see what happens to the rest of the implementation.  Most of
+it is unchanged, but in the `append` method, we now need to account
```

Comment:
> Most of it is unchanged

I assume this implies that the only change is below, in `append`. If that's true, we need to change `count` above (line 283) to return `xs.count`, right? If we're not truncating `ys` in `init` (which, by the way, why?) I think we have to use `xs` as the definitive length for the array.

Reply:
Maybe we can just change the implementation as I suggested, and remove the note altogether? Seems cleaner.